feat(Core): Add unique parameter for accurate replica counting

mgajek-cern opened the pull request:
Closes #8027

---

# Manual Testing Evidence for PR

## Setup

Create and checkout a local tracking branch:
```sh
git remote add mgajek-cern https://github.com/mgajek-cern/rucio.git
# Check upstreams and origin
git remote -v
git fetch mgajek-cern patch-8027-get-local-account-usage-unique
git checkout -b patch-8027-get-local-account-usage-unique mgajek-cern/patch-8027-get-local-account-usage-unique
```

Start the development environment, e.g.:
```bash
docker compose -f etc/docker/dev/docker-compose.yml up -d
docker exec -it dev-rucio-1 bash # Enter container
tools/run_tests.sh -ir
```

## Test Results

**NOTE:** Subsequent commands are invoked within the `dev-rucio-1` container.

**REST API:**
```bash
# Get auth token
TOKEN=$(curl -k -s -D - https://localhost:443/auth/userpass \
  -H "X-Rucio-Account: root" \
  -H "X-Rucio-Username: ddmlab" \
  -H "X-Rucio-Password: secret" \
  | grep "X-Rucio-Auth-Token:" | cut -d' ' -f2 | tr -d '\r')

# Test without unique parameter
curl -k -s "https://localhost:443/accounts/root/usage/local" \
  -H "X-Rucio-Auth-Token: $TOKEN"

# Test with unique=true
curl -k -s "https://localhost:443/accounts/root/usage/local?unique=true" \
  -H "X-Rucio-Auth-Token: $TOKEN"
```

**CLI:**
```bash
rucio list-account-usage root          # Default: counter-based
rucio list-account-usage root --unique # Unique: direct replica query
```

**Automated tests:**
```bash
pytest tests/test_account_limits.py::TestCoreAccountLimitsUnique -v
```

## Observed Results

**Without `--unique`:** Shows 0 bytes (counter-based)
```
XRD1  | 0.000 B
XRD2  | 0.000 B
```

**With `--unique`:** Shows actual usage (direct query)
```
XRD1  | 20.972 MB
XRD2  | 20.972 MB
```

<img width="1032" height="440" alt="image" src="https://github.com/user-attachments/assets/348f3b2e-77e0-43e7-a2f7-b5fc64fe33be" />

This demonstrates the unique parameter correctly queries replicas directly instead of using `cached` counters, accurately reflecting actual storage usage and preventing double-counting when multiple locks exist.

voetberg commented on lib/rucio/cli/bin_legacy/rucio.py:
This functionality also needs to be included in the new cli (and arguably access to it should only be there so we don't introduce new features to a part of code slated for deprecation)

mgajek-cern replied:
Thanks for the feedback. Addressed in subsequent commits

voetberg commented on lib/rucio/client/accountclient.py:
```diff
@@ -457,13 +457,20 @@ def get_local_account_usage(self, account: str, rse: Optional[str] = None) -> "I
             The account name.
         rse :
             The rse name.
+        unique :
```
Would it make sense to include a small warning about this being a computational expensive argument? The caching was done originally to avoid long wait times.

mgajek-cern replied:
Addressed in subsequent commits

voetberg commented on lib/rucio/core/account_limit.py:
```diff
@@ -321,34 +327,34 @@ def get_local_account_usage(account: "InternalAccount", rse_id: Optional[str] =
 
-    for rse_id in set(limits).union(counters):
-        counter = counters.get(rse_id)
+    for rse_id_iter in set(limits).union(counters):
```
Hyper nit-pick comment but why was this name change needed? It seems like an unnecessarily worse name.

mgajek-cern replied:
Addressed in subsequent commits

voetberg commented on lib/rucio/web/rest/flaskapi/v1/accounts.py:
```diff
@@ -917,66 +917,74 @@ def get(self, account: str, rse: str) -> Response:
+      406:
+        description: "Not acceptable"
+    """
+    unique = request.args.get('unique', 'false').lower() == 'true'
```
Other boolean args accept 0 and 1 as false and true
```suggestion
    unique = request.args.get('unique', 'false').lower() in ['true', "1"]
```

mgajek-cern replied:
Addressed in subsequent commits

---

However, according to the [OpenAPI specification](https://swagger.io/docs/specification/v3_0/data-models/data-types/#boolean) for boolean types:

> type: boolean represents two values: true and false. Note that truthy and 
> falsy values such as "true", "", 0 or null are not considered boolean values.

While accepting '1' and '0' maintains consistency with some existing endpoints, 
the OpenAPI spec is clear that booleans should be 'true' or 'false'. Since this 
is a new parameter, we could argue for strict spec compliance here.

voetberg replied:
Thanks for doing the research, if that's the case we should make this consistent across the REST API.

voetberg commented on lib/rucio/web/rest/flaskapi/v1/accounts.py:
```diff
@@ -917,66 +917,74 @@ def get(self, account: str, rse: str) -> Response:
+      schema:
+        type: string
+      style: simple
+    - name: unique
```
This is somewhat inconsistent in the docs (sorry!!!) but the preferred way to denote something is part of the request body and not in the path is to set it separate from the parameters. [Example](https://github.com/rucio/rucio/blob/7086605450c6b6e9a642ca14f6c57729e6fd5fd4/lib/rucio/web/rest/flaskapi/v1/locks.py#L42) 
```
requestBody:
          content:
            application/json:
              schema:
                type: object
                properties:
                  did_type:
                    description: "The DID type to filter for."
                    type: string
                    enum: ['dataset']
```

mgajek-cern replied:
Thanks for the feedback. However, `unique` is a query parameter on a GET request (accessed via `request.args.get()`), not part of a request body. 

According to [RFC 7231 Section 4.3.1](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.1):
> "A payload within a GET request message has no defined semantics; sending a payload body on a GET request might cause some existing implementations to reject the request."

The `requestBody` section is used for POST/PUT/PATCH methods that accept JSON/form data in the body. For GET requests with optional query parameters like `?unique=true`, the current `parameters` section with `in: query` is the correct OpenAPI specification approach.

voetberg replied:
Ah, I misunderstood where this argument was being passed. Okay carry on!

voetberg commented on .github/workflows/integration_tests.yml:
```diff
@@ -148,6 +148,8 @@ jobs:
             pip install --no-cache-dir -e /rucio_source
             tools/run_tests.sh -ir
           "
+      - name: Account Limits Test
```
Can I get a justification for why this needs to be explicitly part of the integration tests instead of just being a part of the rest of the test suite?

mgajek-cern replied:
Enabled it for fast feedback. Removed in subsequent commits

mgajek-cern commented on lib/rucio/cli/account.py:
```diff
@@ -144,13 +144,13 @@ def limit():
+@click.option("--unique", is_flag=True, default=False, help="Count unique replicas to avoid double-counting when multiple locks exist.")
 @click.pass_context
-def limit_list(ctx, account_name, rse):
+def limit_list(ctx, account_name, rse, unique):
```
@voetberg  `autotest.yml` wants type annotations, but the whole CLI module lacks them. Should I add them to this function or is there a reason they're skipped?

```suggestion
def limit_list(ctx: click.Context, account_name: str, rse: Optional[str], unique: bool) -> None:
```

voetberg replied:
Yeah, this is a discussion we need to have in full - see https://github.com/rucio/rucio/issues/7582

haozturk commented:
Thanks @mgajek-cern for this PR. I've tested this by creating multiple locks of the same DID in a given RSE and I confirmed that the unique option works as expected:

1. New account with no data
```
# rucio list-account-usage johnsnow --unique

+-------+---------+---------+--------------+
| RSE   | USAGE   | LIMIT   | QUOTA LEFT   |
|-------+---------+---------+--------------|
+-------+---------+---------+--------------+
+------------------+---------+---------+--------------+
| RSE EXPRESSION   | USAGE   | LIMIT   | QUOTA LEFT   |
|------------------+---------+---------+--------------|
+------------------+---------+---------+--------------+
```
2. Create one lock for file `test:file3` on RSE `XRD1`
```
[root@857e1ccb4a50 rucio]# rucio rule add --copies 1 --rses XRD1 --account johnsnow test:file3
b56eb619398f4341ac7c67e59c87bc44
[root@857e1ccb4a50 rucio]# rucio list-account-usage johnsnow --unique

+-------+-----------+----------+--------------+
| RSE   | USAGE     | LIMIT    | QUOTA LEFT   |
|-------+-----------+----------+--------------|
| XRD1  | 10.486 MB | 1.000 PB | 1000.000 TB  |
| XRD2  | 0.000 B   | 1.000 PB | 1.000 PB     |
+-------+-----------+----------+--------------+
```
3. Create a second lock for the same file on the same RSE:
```
[root@857e1ccb4a50 rucio]# rucio rule add --copies 2 --rses "XRD1|XRD2" --account johnsnow test:file3
7447dd27858d497c96b21440b1ded393
# rucio list-account-usage johnsnow --unique

+-------+-----------+----------+--------------+
| RSE   | USAGE     | LIMIT    | QUOTA LEFT   |
|-------+-----------+----------+--------------|
| XRD1  | 10.486 MB | 1.000 PB | 1000.000 TB  |
| XRD2  | 10.486 MB | 1.000 PB | 1000.000 TB  |
+-------+-----------+----------+--------------+
``` 
As it can be seen, the usage in XRD1 didn't increase from step 2 to 3 as expected

haozturk reviewed (commented):
The expected behavior is achieved according to a simple test I've executed. Haven't reviewed the code.

mgajek-cern commented:
Rebasing onto latest master and squashing 9 commits into a single conventional commit

mgajek-cern commented:
CI is passing, and the initial findings from @voetberg’s previous review have been addressed.
Are any further reviews required, or can we approve and merge this non-breaking change?

Geogouz commented on lib/rucio/core/account_limit.py:
```diff
@@ -401,3 +407,79 @@ def get_global_account_usage(account: "InternalAccount", rse_expression: Optiona
+                'bytes': usage['bytes'],
+                'files': usage['files'],
+                'bytes_limit': limit,
+                'bytes_remaining': limit - usage['bytes'] if limit > 0 else 0,
```
Whenever no quota is defined (limit defaults to 0), this reports `bytes_remaining` as 0, even if the account already consumes space. However, the pre-existing counter-based path returns negative values in that case (0 - usage), which I assume is how over-quota situations without an explicit limit are surfaced today. So consumers requesting `unique=True` would miss over-consumption because they would never see negative values there. I think we should preserve the original calculation (limit - usage) regardless of limit value.

mgajek-cern replied:
Indeed. This preserves RUCIO's established negative quota pattern (indicating over-consumption) 
for consistency with existing APIs, though some systems (AWS, K8s) floor quotas at 0.
Implicitly stated here ( "percentage of quota in use" implies 0-100%): https://docs.aws.amazon.com/servicequotas/latest/userguide/intro.html 

Rebased and removed the conditional expression in subsequent commit.

Geogouz commented:
Some `ruff` checks to be fixed pending.

mgajek-cern commented:
> Some `ruff` checks to be fixed pending.

Addressed previous CI findings and rebased onto upstream/master

voetberg commented on lib/rucio/client/accountclient.py:
```diff
@@ -458,13 +458,23 @@ def get_local_account_usage(self, account: str, rse: Optional[str] = None) -> "I
         rse :
             The rse name.
+        unique :
+            unique :
```
Unique is listed as an arg twice

mgajek-cern replied:
Addressed

voetberg commented on lib/rucio/cli/account.py:
```diff
@@ -144,13 +146,13 @@ def limit():
 @limit.command("list", help="Shows the space used, the quota limit and the quota left for an account for every RSE where the user have quota.")
 @click.argument("account-name")
 @click.option("--rse", "--rse-name", help="Show usage for only for this RSE.")
+@click.option("--unique", is_flag=True, default=False, help="Count unique replicas to avoid double-counting when multiple locks exist.")
```
Should the warning on the Client be duplicated here?

mgajek-cern replied:
Addressed

dchristidis commented on lib/rucio/core/account_limit.py:
```diff
@@ -321,11 +327,12 @@ def get_local_account_usage(account: "InternalAccount", rse_id: Optional[str] =
     else:
         # One RSE
-        stmt.where(
+        stmt = stmt.where(
```
I have the impression you found a fix for #7125. Please split it into a separate PR and reference that issue accordingly.

mgajek-cern replied:
The #7125 bug (RSE filter not working in get_local_account_usage) exists in the counter-based path and is independent of this PR's unique parameter. This PR doesn't address that filtering issue. I can create a separate PR to fix #7125 if needed.

dchristidis replied:
I should rephrase: I am sure this on-line change fixes the aforementioned issue.

mgajek-cern replied:
Reverted changes that fix #7125

dchristidis replied:
Ideally, if you could open that other PR at your earliest convenience then I can merge it immediately, and then you can rebase this PR on top of the (future) master.

mgajek-cern replied:
PR has been created https://github.com/rucio/rucio/pull/8336

dchristidis replied:
Superb, thanks for your cooperation.

dchristidis commented on lib/rucio/core/account_limit.py:
```diff
@@ -348,7 +355,6 @@ def get_local_account_usage(account: "InternalAccount", rse_id: Optional[str] =
             })
     return result_list
 
-
```
Why? Two lines follow [PEP 8](https://peps.python.org/pep-0008/#blank-lines).

mgajek-cern replied:
Addressed

dchristidis commented on lib/rucio/core/account_limit.py:
```diff
@@ -401,3 +407,79 @@ def get_global_account_usage(account: "InternalAccount", rse_expression: Optiona
+        models.RSEFileAssociation
+    ).join(
+        models.ReplicaLock,
+        and_(
+            models.ReplicaLock.scope == models.RSEFileAssociation.scope,
+            models.ReplicaLock.name == models.RSEFileAssociation.name,
+            models.ReplicaLock.rse_id == models.RSEFileAssociation.rse_id
+        )
+    ).where(
+        models.ReplicaLock.account == account
+    ).distinct()  # DISTINCT on the entire row (scope, name, rse_id, bytes)
```
Firstly, I don’t understand the need to go through the replicas table. Why not go over the locks directly?

Secondly, I have some concerns as to whether this will work in practice for large instances. I think only with an index on the account, but this requires some discussion.

mgajek-cern replied:
Good point. I used replicas as the authoritative source for bytes, though ReplicaLock also has it as a denormalized copy. Both approaches should produce the same result for active locks.
Could switch to locks-only if there are clear performance advantages. 

What query volume are you expecting, and any indexing suggestions?

dchristidis replied:
I’m working under the assumption that, when going over two tables that require a full table scan, going over just one of them will more efficient.

The locks table can become very large. For a testing environment it won’t show, but for ATLAS (and assume so for CMS as well) a full table scan during a request on the servers may prove unacceptable.

The index could be on the account column (e.g. adding something like `Index('LOCKS_ACCOUNT_IDX', 'account')` to the model plus the migration changes), but I defer to @bari12 to say what is the right thing to do here.

mgajek-cern replied:
Thanks for clarifying. You're right that a full table scan on locks would be problematic at scale. An index on locks.account makes sense. I can change to a locks-only query. Let me wait for @bari12's input before updating.

mgajek-cern replied:
Updated to locks-only approach in subsequent commit. The query now uses ReplicaLock directly without joining RSEFileAssociation, avoiding full table scans on two large tables. This addresses the performance concern for large-scale instances.

---

Regarding indexing on locks.account: I can add the index in a follow-up PR once we confirm this is the preferred approach or can include it in this PR or handle separately.

dchristidis replied:
> This addresses the performance concern for large-scale instances.

Not exactly. The concern remains and only, in my estimate, the addition of an index would make this safe. But the query is now in line with my original expectations.

mgajek-cern replied:
I've added the index on locks.account to address the performance concern.
Question: The old code was already joining on locks (filtering by RSE). Was that not causing performance issues or is the missing index a longstanding problem that my changes just made more visible?

dchristidis replied:
> I've added the index on locks.account to address the performance concern.

Please also add it to `etc/sql/oracle/schema.sql`.

> The old code was already joining on locks (filtering by RSE).

Forgive me, which code are you referring to?

mgajek-cern replied:
> Please also add it to `etc/sql/oracle/schema.sql`.

Added to `etc/sql/oracle/schema.sql`.

> Forgive me, which code are you referring to?

I was referring to [this outerjoin at line 304 in lock.py](https://github.com/rucio/rucio/blob/master/lib/rucio/core/lock.py#L304) that joins DataIdentifierAssociation with ReplicaLock filtering by RSE. However, I now realize that query likely performs correctly because it filters by rse_id, which would have already has an index (either through foreign key constraints or explicit indexing).

dchristidis replied:
What, I believe, plays a more important role is that in that query the left side (i.e. the content of a dataset) is expected to be small whereas in your case (i.e. the replicas of an RSE) it is expected to be very large.

mgajek-cern replied:
Thank you for clarifying. I see now that the result set size difference (small dataset contents vs. potentially many account locks) is the more critical factor here.

mgajek-cern commented:
Squashed 5 commits into a single commit per the [contributing guideline](https://rucio.cern.ch/documentation/contributing/#conventional-commits): "Using multiple commits is allowed as long as they achieve an independent, well-defined, change and are well-described. Otherwise multiple commits should be squashed."
All changes implement the same feature (#8027), so consolidating into one commit provides clearer git history.

dchristidis commented:
> All changes implement the same feature

The inclusion of a new index is significant enough that it warrants a separate commit.

mgajek-cern commented:
> The inclusion of a new index is significant enough that it warrants a separate commit.

We got 1 commit on Database schema change and 1 commit on Feature implementation now

dchristidis commented on lib/rucio/db/sqla/migrate_repo/versions/027f78eff9d9_add_index_on_locks_account.py:
```diff
@@ -0,0 +1,49 @@
+    if dialect_name == 'mysql':
+        # MySQL automatically creates an index for foreign key columns
+        # Check if index already exists before creating
+        pass
```
As far as I can see, no other migration does this. Could you explain (1) if it is necessary and (2) if it is useful to have this conditional?

mgajek-cern replied:
Good catch. You're right it's not common in other migrations.

**Why it's here:** MySQL auto-creates indexes for FK columns, so we'd create a duplicate if we didn't skip it. The explicit `pass` with comment documents this behavior so future developers don't try to "fix" it by adding the index.

**Is it necessary?** Technically no. We could just use `if dialect_name in ['oracle', 'postgresql']` without the MySQL branch.

**Is it useful?** I think yes. it makes the intent explicit and prevents someone from later adding MySQL "by mistake." But happy to remove it if you prefer consistency with other migrations.

dchristidis replied:
> MySQL auto-creates indexes for FK columns, so we'd create a duplicate if we didn't skip it.

This is the part that I’m trying to understand, though I admit I did not spend any effort researching myself. What are the implications of ‘duplicate’? Is it two separate indexes, on implicit and one explicit, taking twice the volume on the storage?

mgajek-cern replied:
Removed the MySQL branch for consistency with other migrations.

Regarding duplicate indexes: InnoDB automatically creates an index on foreign key columns in the referencing (child) table if one does not already exist (per [MySQL documentation](https://dev.mysql.com/doc/refman/8.0/en/create-table-foreign-keys.html#foreign-key-indexes)). Explicitly creating an index on the same columns is therefore usually redundant and can increase storage usage and write overhead. While InnoDB may later drop a redundant index if a suitable one exists, other migrations do not handle this conditionally, so the MySQL branch was removed for consistency.

dchristidis replied:
> so the MySQL branch was removed for consistency.

As far as I know, no community uses MySQL in production, so the simplification of our migration scripts seems beneficial.

mgajek-cern committed:
feat(Database)!: Add index on locks.account for performance

Add index on ReplicaLock.account column to optimize account-based
lock queries at ATLAS/CMS scale. Prevents full table scans when
querying locks by account.

BREAKING CHANGE: This change requires running database migration
027f78eff9d9 which must be executed manually outside the typical
Rucio deployment process for Oracle and PostgreSQL databases.
MySQL is not affected as it auto-creates indexes on foreign keys.

Issue: #8027

mgajek-cern commented:
This PR has been open for ~5 months and has gone through multiple review rounds, refactors and rebases. All currently raised concerns appear to be addressed and CI is green.

@dchristidis - your review is currently blocking the merge. Before spending more time iterating, I’d like to ask for a clear decision: is this PR ready to merge or are there blocking issues that would prevent merging it in its current form? A definitive answer either way would be appreciated so we can close the loop.

dchristidis commented on lib/rucio/db/sqla/migrate_repo/versions/027f78eff9d9_add_index_on_locks_account.py:
```diff
@@ -0,0 +1,40 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""add_index_on_locks_account"""    # noqa: D`400, D415
```
Is the backtick here on purpose?

mgajek-cern replied:
Addressed nit in subsequent commit

dchristidis replied:
Please just remove the backtick and keep everything else as before. Meaning:

```
"""add_index_on_locks_account"""    # noqa: D400, D415
```

The use of single quotes for triple-quoted docstrings goes against PEP 257.

mgajek-cern replied:
Just wanted to align with/mirrored whats been done in:

- 1a80adff031a_create_index_on_rules_hist_recent.py
- 1c45d9730ca6_increase_identity_length.py
- 1a80adff031a_create_index_on_rules_hist_recent.py
- 0f1adb7a599a_create_transfer_hops_table.py

and the vast majority of migration files

mgajek-cern replied:
> Please just remove the backtick and keep everything else as before. Meaning:
> 
> ```
> """add_index_on_locks_account"""    # noqa: D400, D415
> ```
> 
> The use of single quotes for triple-quoted docstrings goes against PEP 257.

All good now. Would be good to enforce formatting concerns via a tool if possible in the future

dchristidis commented on lib/rucio/db/sqla/migrate_repo/versions/027f78eff9d9_add_index_on_locks_account.py:
```diff
@@ -0,0 +1,40 @@
+
+
+def upgrade():
+    """Upgrade schema."""
```
Why the deviation from the template for the two docstrings?

mgajek-cern replied:
Addressed nit in subsequent commit

dchristidis replied:
I repeat, why the deviation from the template?

https://github.com/rucio/rucio/blob/e9a5a52fbb5e5b9439eeaea57e691ba08756e68e/lib/rucio/db/sqla/migrate_repo/script.py.mako#L27

mgajek-cern replied:
Just wanted to align with/mirrored whats been done in:

- 1a80adff031a_create_index_on_rules_hist_recent.py
- 1c45d9730ca6_increase_identity_length.py
- 1a80adff031a_create_index_on_rules_hist_recent.py
- 0f1adb7a599a_create_transfer_hops_table.py

and the vast majority of migration files

dchristidis replied:
Past migration scripts used a previous version of the template. It was deemed unnecessary to update them. Didn’t you start with `alembic revision`? It would have created the base of the migration, based on the template. All these deviations take extra effort.

mgajek-cern replied:
All good now

dchristidis commented on lib/rucio/db/sqla/migrate_repo/versions/027f78eff9d9_add_index_on_locks_account.py:
```diff
@@ -0,0 +1,40 @@
+    dialect_name = context.get_context().dialect.name
+    schema = context.get_context().version_table_schema if context.get_context().version_table_schema else None
+
+    if dialect_name in ['oracle', 'postgresql']:
```
Isn’t the lack of `mysql` here introducing a bit of an inconsistency between the model (where `LOCKS_ACCOUNT_IDX` is expected to exist) and the migration script (which doesn’t create an index with such a name)?

mgajek-cern replied:
The MySQL removal was agreed upon previously.

See: https://github.com/rucio/rucio/pull/8072#discussion_r2711922008

dchristidis replied:
Not from my part. Apologies if I did not make myself understood.

The original comment reads:

> As far as I can see, no other migration does this. Could you explain (1) if it is necessary and (2) if it is useful to have this conditional?

I did not mean that you should remove the conditional part that skipped MySQL, but that it should be treated exactly like the other databases.

mgajek-cern replied:
Appreciate the consistency concern, but there's a technical constraint: adding MySQL to this migration causes CI failures because InnoDB automatically creates indexes on foreign key columns. Trying to explicitly create LOCKS_ACCOUNT_IDX results in duplicate index errors.
See CI run demonstrating the failure: https://github.com/mgajek-cern/rucio/actions/runs/21129412652/job/60756980850#step:7:2719 (`pymysql.err.OperationalError: (1061, "Duplicate key name 'LOCKS_ACCOUNT_IDX'")`)
Given this, the current approach (excluding MySQL from the migration while keeping it in the model) is the pragmatic choice:

- No CI failures
- MySQL still gets the index (via InnoDB auto-creation)
- Consistent with production database focus (Oracle/PostgreSQL)

This aligns with your earlier comment that [no community uses MySQL in production](https://github.com/rucio/rucio/pull/8072#discussion_r2731196582), and the expectation from yesterday's meeting that MySQL will phase out at some point.

dchristidis replied:
This is a very useful finding, thank you. Allow me look closer and get back to you.

mgajek-cern replied:
Thanks again for the sync @dchristidis. To sum up this concern after today’s meeting:

MySQL/InnoDB automatically creates an index on foreign key columns if one does not already exist (MySQL 8.0 Reference Manual, FOREIGN KEY Constraints → Conditions and Restrictions).
When the migration explicitly attempts to create LOCKS_ACCOUNT_IDX, MySQL raises ER_DUP_KEYNAME (1061) because InnoDB has already created the necessary index for foreign key enforcement.
This is expected behavior. Explicitly creating the same index is redundant and unsafe and the resulting CI failure is consistent with MySQL’s documented behavior.

Ref: 

- https://github.com/rucio/rucio/pull/8072#discussion_r2740980680
- https://github.com/rucio/rucio/pull/8072#discussion_r2711922008

---
The PR should now be in a mergeable state

dchristidis commented on lib/rucio/cli/account.py:
```diff
@@ -144,13 +146,20 @@ def limit():
+    args = Arguments({"no_pager": ctx.obj.no_pager, "usage_account": account_name, "rse": rse, "unique": unique})
     list_account_usage(args, ctx.obj.client, ctx.obj.logger, ctx.obj.console, ctx.obj.spinner)
 
-
```
Why?

dchristidis replied:
Related to #8362.

mgajek-cern replied:
Solved after rebasing latest changes in upstream/master

dchristidis replied:
Still there.

mgajek-cern replied:
Can you check again? This should be solved in latest revision

dchristidis commented on lib/rucio/cli/bin_legacy/rucio.py:
```diff
@@ -1811,7 +1811,12 @@ def list_account_usage(args, client, logger, console, spinner):
         spinner.start()
 
-    usage = client.get_local_account_usage(account=args.usage_account, rse=args.rse)
+    unique = getattr(args, 'unique', False)
```
Under what circumstances would this be necessary, given that Click defaults it to `False`?

mgajek-cern replied:
Addressed nit in subsequent commit

dchristidis replied:
But then there’s no reason to assign it to the variable at all. You can use it as a keyword argument, like it is done for the others.

mgajek-cern replied:
Addressed

dchristidis commented on lib/rucio/web/rest/flaskapi/v1/accounts.py:
```diff
@@ -918,66 +918,74 @@ def get(self, account: str, rse: str) -> Response:
-        except AccessDenied as error:
-            return generate_http_error_flask(401, error)
+  @check_accept_header_wrapper_flask(['application/x-json-stream'])
+  def get(self, account: str, rse: Optional[str] = None) -> Response:
```
The indentation in this part uses two spaces instead of four.

mgajek-cern replied:
This indentation issue exists in the current codebase and is unrelated to the `unique` parameter feature. We can create a separate PR to fix formatting issues across the file to keep this PR focused.

dchristidis replied:
https://github.com/rucio/rucio/blob/e9a5a52fbb5e5b9439eeaea57e691ba08756e68e/lib/rucio/web/rest/flaskapi/v1/accounts.py#L919-L922

I don’t see how.

mgajek-cern replied:
Addressed in subsequent commit

dchristidis replied:
Please double check your work. The docstring and the body of the function still use improper indentation.

mgajek-cern replied:
Addressed now. Should enforce this via a formatting tool ideally in the future

dchristidis replied:
Indentation level is also related to #8362. The lack of enforcement is a recent regression.

dchristidis commented on lib/rucio/web/rest/flaskapi/v1/accounts.py:
```diff
@@ -918,66 +918,74 @@ def get(self, account: str, rse: str) -> Response:
+      406:
+        description: "Not acceptable"
+    """
+    unique = request.args.get('unique', 'false').lower() in ['true', '1']
```
Why not use `param_get_bool()` here?

mgajek-cern replied:
Addressed nit in subsequent commit

dchristidis reviewed (changes requested):
Please also add a `BREAKING CHANGE` to the first commit since it requires a migration which is done outside the typical Rucio deployment.

mgajek-cern commented:
> Please also add a `BREAKING CHANGE` to the first commit since it requires a migration which is done outside the typical Rucio deployment.

Done in 538c025. The migration commit now includes the ! marker and a BREAKING CHANGE footer clarifying that 027f78eff9d9 must be run manually for Oracle/PostgreSQL deployments.

NOTE on scope: The original issue [explicitly acknowledged](https://github.com/rucio/rucio/issues/8027#issuecomment-3306131123) this would be a direct database query with acceptable performance trade-offs for opt-in usage. The index requirement emerged during review to address production-scale concerns for ATLAS/CMS deployments, which is a reasonable enhancement but represents an evolution of the original requirements.

mgajek-cern committed:
feat(Core): Add unique parameter for accurate replica counting

Add optional unique parameter to get_local_account_usage that queries
locks directly with DISTINCT to avoid double-counting when multiple
locks exist for the same replica.

Closes: #8027

dchristidis commented:
When the reviewer requests corrections, you are supposed to amend your existing commits where possible, per the [contributing guide](https://rucio.cern.ch/documentation/contributing/#6-watch-the-pull-request-for-reviews).

mgajek-cern committed:
style(Core): Address code review feedback on replica counting feature

- Remove unnecessary getattr() in bin_legacy/rucio.py (Click defaults unique to False)
- Fix migration docstring formatting to match template style
- Use param_get_bool() helper instead of manual parsing in Flask API
- Use 4 space indentation

Issue: #8027

mgajek-cern committed:
test(Testing): Add CLI test for unique parameter in account limit list
Validates that the --unique flag is accepted by the account limit list
command without errors. In test environments without replica data, both
default and unique outputs may be identical; this test ensures the CLI
interface works correctly.

Issue: #8027

mgajek-cern commented:
Let’s close this PR for now and take a more structured approach to the problem.

Given the scope and the feedback so far, it would be better to break this into smaller, focused sub-issues, each addressed by its own PR. This should make reviews easier and reduce the risk of introducing unintended complexity.

We can also consider using stacked PRs to keep related changes connected while still keeping each step reviewable. See comment https://github.com/rucio/rucio/pull/8081#issuecomment-4344190716
